execute/transport.go

```diff
@@ -423,6 +432,60 @@ func (m *finishMsg) Dup() Message {
 	return m
 }
 
+type ProcessChunkMsg interface {
+	Message
+	TableChunk() table.Chunk
+}
+
+type processChunkMsg struct {
+	srcMessage
+	chunk table.Chunk
+	acked bool
+}
+
+func (m *processChunkMsg) Type() MessageType {
+	return ProcessChunkType
+}
+func (m *processChunkMsg) TableChunk() table.Chunk {
+	return m.chunk
+}
+func (m *processChunkMsg) Ack() {
+	if !m.acked {
+		m.chunk.Release()
+		m.chunk = table.Chunk{}
+		m.acked = true
+	}
+}
```

Comment:
My view of it is that when you construct the message you transfer ownership to the message.

The rest is correct. A message will own a reference to whatever is contained inside of it and, when acked, will release that message.

Reply:
So is it on the client code (e.g., the transformations) to call `Retain` so that the message that gets created has ownership? Wouldn't it be better to bundle message creation with a call to `Retain` in a single function (like `NewProcesChunkMsg`) so that we eliminate the danger of client code forgetting to do it?

Reply:
In general, I would think that the ownership gets transferred so there's no need to invoke `Retain`. Usually what happens is the transformation creates a new table chunk and it would wrap that table chunk in the message and send it off. If the client code wanted to keep a reference to it, then a `Retain` would be needed to make a "copy" of the table.

In general, I think this is probably a better question for when we #3887 since that's going to be responsible for constructing the messages.